---
title: "16S_Amplicon_Analysis_Workshop"
author: "Jinny Yang"
date: "2025-08-19"
output:
  html_document: default
---

# Install packages
```{r install-packages, message=FALSE, warning=FALSE, echo=FALSE}
install.packages(
  c(
    "ggh4x",
    "ggplot2",
    "tidyverse",
    "vegan",
    "dplyr",
    "indicspecies",
    "VennDiagram",
    "ape",
    "ggpubr",
    "car",
    "picante",
    "reshape2"
  ),
  repos = "https://cloud.r-project.org"
)

BiocManager::install(c("phyloseq", "DESeq2"))
```

# Load packages
```{r load-packages, message=FALSE, warning=FALSE}

library(ggh4x)
library(phyloseq)
library(ggplot2)
library(tidyverse)
library(vegan)
library(dplyr)
library(indicspecies)
library(VennDiagram)
library(ape)
library(ggpubr)
library(DESeq2)
library(car)
library(picante)
library(reshape2)
```
# Data preparation
## Make metadata from sample names

### ***"feature-table.tsv" was manually edited to replace "#OTU ID" with "ASVID" and named as "feature-table_New.tsv"
```{r make-metadata}
ASVTable <- read.table("feature-table_New.tsv", header = TRUE, sep = "\t", row.names = 1)

P_con <- rep("Initial", ncol(ASVTable))
P_con[which(grepl("625", colnames(ASVTable)))] <- "625uM"
P_con[which(grepl("50", colnames(ASVTable)))] <- "50uM"

Treatment <- rep("NA", ncol(ASVTable))
Treatment[which(grepl("A", colnames(ASVTable)))] <- "Invasion"
Treatment[which(grepl("SCR", colnames(ASVTable)))] <- "Initial SynCom"
Treatment[which(grepl("Soil", colnames(ASVTable)))] <- "Initial SoilMicrobes"

MetaData <- data.frame(P_con, Treatment)
rownames(MetaData) <- colnames(ASVTable)
MetaData
```

## Prepare taxonomy

```{r prepare-taxonomy}
Taxaonomy <- read.csv("taxonomy.tsv", header = TRUE, sep = "\t", row.names = 1)
TAX <- data.frame(
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 1),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 2),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 3),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 4),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 5),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 6),
  sapply(strsplit(Taxaonomy[,1], ";", fixed = TRUE), "[", 7)
)
rownames(TAX) <- rownames(Taxaonomy)
colnames(TAX) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
write.csv(TAX, "Taxonomy.csv")
head(TAX)
```

## Input files to phyloseq
```{r input-phyloseq}
Tree <- read.tree("tree.nwk")

physeq <- phyloseq(
  otu_table(ASVTable, taxa_are_rows = TRUE),
  tax_table(as.matrix(TAX)),
  sample_data(MetaData),
  phy_tree(Tree)
)
```

## Filter taxonomy: remove mitochondria and chloroplast, and keep only "Bacteria"
```{r filter-taxonomy}
physeq <- physeq %>%
  subset_taxa(
    Family   != "f__Chloroplast" &
    Family   != "f__Mitochondria" &
    Kingdom  == "d__Bacteria"
  )
```

##### The phyloseq package is ready for use!! Let's go some analysis!!

## Analyses don't need rarefaction (Sequencing depth is taken account during analysis) 

## 1. Beta diversity
### Visualization: Ordination analysis (Bray-Curtis/Unifrac or PCoA/NMDS)
##### Question: Does community composition differ between different groups (Treatment and P_con)?
```{r beta-diversity-ordination, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=4}
#pdf("Ordination.pdf", width=8, height=4)

Figure1 <- ggarrange(
  plot_ordination(physeq, ordinate(physeq, "PCoA", distance = "bray"), "samples", color = "Treatment", shape = "P_con") + ggtitle("BC x PCoA"),
  plot_ordination(physeq, ordinate(physeq, "PCoA", distance = "unifrac"), "samples", color = "Treatment", shape = "P_con") + ggtitle("Unifrac x PCoA"),
  plot_ordination(physeq, ordinate(physeq, "NMDS", distance = "bray"), "samples", color = "Treatment", shape = "P_con") + ggtitle("BC x NMDS"),
  plot_ordination(physeq, ordinate(physeq, "NMDS", distance = "unifrac"), "samples", color = "Treatment", shape = "P_con") + ggtitle("Unifrac x NMDS"),
  common.legend = TRUE
)

Figure1

#dev.off()
```

###### *The example data set may be too simple and not fit for NMDS analysis

### Statistical analysis (using Bray-Curtis distance as an example):
#### PERMANOVA (Between-group centroids, mean shifts)
##### Question: Are community compositions significantly different between groups?

```{r beta-diversity-permanova, message=FALSE, warning=FALSE}
BC_dist <- phyloseq::distance(physeq, method = "bray")

set.seed(123)

BC_PERMANOVA_Interaction <- adonis2(
  formula = BC_dist ~ Treatment * P_con,
  permutations = 1000,
  data = data.frame(sample_data(physeq))
)
BC_PERMANOVA_Interaction

BC_PERMANOVA_Treatment <- adonis2(
  formula = BC_dist ~ Treatment,
  permutations = 1000,
  data = data.frame(sample_data(physeq))
)
BC_PERMANOVA_Treatment

BC_PERMANOVA_P <- adonis2(
  formula = BC_dist ~ P_con,
  permutations = 1000,
  data = data.frame(sample_data(physeq))
)
BC_PERMANOVA_P
```

#### ANOSIM (Between-group vs within-group ranks, distance to average distance)
##### Question: Does communites between groups are more different then within group?

```{r beta-diversity-anosim, message=FALSE, warning=FALSE}
set.seed(123)

anosim_Treatment <- anosim(BC_dist, data.frame(sample_data(physeq))$Treatment, permutations = 1000)
anosim_Treatment

anosim_P <- anosim(BC_dist, data.frame(sample_data(physeq))$P_con, permutations = 1000)
anosim_P
```

#### betadisper (Within-group dispersion, distance to centroids)
##### Question: Does communities within the same group non-significantly different?
```{r beta-diversity-betadisper, message=FALSE, warning=FALSE,fig.width=4, fig.height=4}
BetaDis_Treatment <- betadisper(BC_dist, group = sample_data(physeq)$Treatment)
BetaDis_P <- betadisper(BC_dist, group = sample_data(physeq)$P_con)
anova(BetaDis_Treatment)
anova(BetaDis_P)

plot(BetaDis_Treatment, label = FALSE)
plot(BetaDis_P, label = TRUE, label.cex = 0.5)
```


#### Differential abundance analysis using DESeq2
##### Question: How many ASV show differential abundance between high and low Pi condition?
```{r differential-abundance-DESeq2, message=FALSE, warning=FALSE}
physeq_subset = subset_samples(physeq, Treatment %in% c("Invasion"))

dds <- phyloseq_to_deseq2(physeq_subset, ~ P_con)
dds <- estimateSizeFactors(dds, type = "poscounts") 
dds <- DESeq(dds, test = "Wald", fitType = "parametric")
res <- results(dds, cooksCutoff = FALSE)

SigDE <- res[which(res$padj <= 0.05), ]
SigDE
```
#####▫ Negative value: The ASV is less abundant in 625 uM than in 50 uM.

#####▫ Positive value: The ASV is more abundant in 625 uM than in 50 uM.

## Analyses need rarefaction
#### Rarefaction
```{r rerafaction, message=FALSE, warning=FALSE}
min(sample_sums(physeq))

rarecurve <- rarecurve(t(as(otu_table(physeq), "matrix")), step = 100, xlab = "Sample Size", ylab = "ASV", label = FALSE, tidy = TRUE)

ggplot(rarecurve, aes(Sample, Species)) + 
  geom_point(size=0.1) + 
  labs(title="16S rarefaction curves", x="16S reads per sample") + 
  geom_vline(xintercept = min(sample_sums(physeq)), col="red")

rare_physeq = rarefy_even_depth(physeq, rngseed=123, sample.size=min(sample_sums(physeq)))
write.csv(otu_table(rare_physeq),"Rarefied_17900_ASVTable.csv")
```

#### Alpha diversity, richness and phylogenetic diversity
##### Questions: does higher Pi harbor more diversity of bacteria under invaion condition?
##### Start with a quick peek
```{r alpha-diversity-peek!, message=FALSE, warning=FALSE}
rare_physeq_subset <- subset_samples(rare_physeq, Treatment %in% c("Invasion"))

plot_richness(rare_physeq_subset, x="P_con", measures=c("Observed","Shannon","InvSimpson"))
```

##### Significantly different? Statistical analysis!
```{r alpha-diversity_statisitc, message=FALSE, warning=FALSE}
AlphaD <- estimate_richness(rare_physeq_subset)
# calcualte phylogenetic diversity (PD)
AlphaPD <- pd(t(as(otu_table(rare_physeq_subset), "matrix")), phy_tree(rare_physeq_subset), include.root = TRUE)
# combine them with meta data
AlphaD_meta <- data.frame(sample_data(rare_physeq_subset),AlphaD,AlphaPD)

## If groups variance different
leveneTest(Observed ~ P_con, AlphaD_meta)
leveneTest(Shannon ~ P_con, AlphaD_meta)
leveneTest(Shannon ~ P_con, AlphaD_meta)
leveneTest(PD ~ P_con, AlphaD_meta)

#If groups mean different
anova_observed <- aov(Observed ~ P_con, data=AlphaD_meta)
summary(anova_observed)
anova_Shannon <- aov(Shannon ~ P_con, data=AlphaD_meta)
summary(anova_Shannon)
anova_InvSimpson <- aov(InvSimpson ~ P_con, data=AlphaD_meta)
summary(anova_InvSimpson)
anova_PD <- aov(PD ~ P_con, data=AlphaD_meta)
summary(anova_PD)
```

###### None of them are significantly different...:(


#### Taxonomy stack bar plots
##### Question: How does taxonomic composition looks like? (Phylum level for example)
```{r taxonomy, message=FALSE, warning=FALSE}
rare_physeq_phylum <- tax_glom(rare_physeq, taxrank = "Phylum")
rare_physeq_phylum_melted = psmelt(rare_physeq_phylum)

rare_physeq_phylum_melted$Phylum[rare_physeq_phylum_melted$Abundance/17900 < 0.01] <- "<1%" 
#ASV tables were rarefied at 17900

ggplot(rare_physeq_phylum_melted, aes(x = Sample, y = Abundance/17900, fill = Phylum)) +
  geom_bar(stat = "identity") +
  facet_wrap2(Treatment~., scales = "free_x") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab("Relative Abundance") +
  labs(fill = "Phylum")+ xlab("")
```

#### Rank abundance analysis on SynCom ASV
##### Does SynCom colonization change under different Pi supply levels?? (Consider SynCom members only)
```{r RDA_SynCom, message=FALSE, warning=FALSE,fig.width=8, fig.height=7}
Rare_table <- otu_table(rare_physeq)
Rare_taxa <- as.data.frame(tax_table(rare_physeq))
Blast_mannual <- read.table("SynComASV_results_97_unique.txt",row.names = 1,header=T)

df1_50 <- Rare_table[, 1:3] %>%
  as.data.frame() %>%
  rownames_to_column("Row.names") %>%
  mutate(mean = rowMeans(select(., -Row.names), na.rm = TRUE)) %>%
  select(Row.names, mean)

df1_625 <- Rare_table[, 4:6] %>%
  as.data.frame() %>%
  rownames_to_column("Row.names") %>%
  mutate(mean = rowMeans(select(., -Row.names), na.rm = TRUE)) %>%
  select(Row.names, mean)

df2 <- Blast_mannual["sseqid"] %>% as.data.frame() %>% rownames_to_column("Row.names")
df3 <- Rare_taxa["Order", drop = FALSE] %>% as.data.frame() %>% rownames_to_column("Row.names")

P50 <- df1_50 %>%
  inner_join(df2, by = "Row.names") %>%
  inner_join(df3, by = "Row.names") %>%
  group_by(sseqid, Order) %>%
  summarize(mean = sum(mean, na.rm = TRUE), .groups = "drop")

P625 <- df1_625 %>%
  inner_join(df2, by = "Row.names") %>%
  inner_join(df3, by = "Row.names") %>%
  group_by(sseqid, Order) %>%
  summarize(mean = sum(mean, na.rm = TRUE), .groups = "drop")

A <- ggplot(P50, aes(x = reorder(sseqid, -mean), y = mean/17900, fill = Order)) +
  geom_bar(stat = "identity") +
  labs(
    x = "SynCom",
    y = "Mean relative abundance",
    title = "Rank Abundance Distribution: 50uM P",
    fill = "Order"
  ) +ylim(0,0.415)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

B <- ggplot(P625, aes(x = reorder(sseqid, -mean), y = mean/17900, fill = Order)) +
  geom_bar(stat = "identity") +
  labs(
    x = "SynCom",
    y = "Mean relative abundance",
    title = "Rank Abundance Distribution: 625uM P",
    fill = "Order"
  ) +ylim(0,0.415)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

ggarrange(A,B,common.legend=T, nrow=2)

```

### To make your plots prettier: https://rfortherestofus.com/2025/04/ggplot2-theme

